src/H5private.h

```diff
@@ -1264,17 +1270,24 @@ H5_DLL herr_t H5CX_pop(bool update_dxpl_props);
                             FUNC_ENTER_COMMON_NOERR(H5_IS_API(__func__));                                    \
                             {
 
+/* Note: this macro only works when there's _no_ interface initialization routine for the module */
+#define FUNC_ENTER_NOAPI_INIT(err)                                                                           \
+    /* Initialize the package, if appropriate */                                                             \
+    H5_PACKAGE_INIT(H5_MY_PKG_INIT, err)
+
 /* Use this macro for all "normal" non-API functions */
 #define FUNC_ENTER_NOAPI(err)                                                                                \
     {                                                                                                        \
         FUNC_ENTER_COMMON(!H5_IS_API(__func__));                                                             \
-        {
+        FUNC_ENTER_NOAPI_INIT(err)                                                                           \
+        if (H5_LIKELY(H5_PKG_INIT_VAR || !H5_TERM_GLOBAL)) {
 
 /* Use this macro for all non-API functions, which propagate errors, but don't issue them */
 #define FUNC_ENTER_NOAPI_NOERR                                                                               \
     {                                                                                                        \
         FUNC_ENTER_COMMON_NOERR(!H5_IS_API(__func__));                                                       \
-        {
+        FUNC_ENTER_NOAPI_INIT(-)                                                                             \
```

Comment:
Why the dash here? This eventually gets passed to `HGOTO_ERROR` as the failure value, and this doesn't seem to be a valid value.

Reply:
Yes, it's designed to be an invalid value that will cause a compile error.   The underlying macro shouldn't get expanded with the macro parameter, so passing a '-' guarantees a compile error if something changes in the code to make the macro expand.